### ISSUE TITLE
Rate limit is sometimes ignored

### ISSUE THREAD

**Issue by huxcrux** [zmap/zmap#903]
> Notice: Please don't file a bug report for questions on how to use ZMap, etc. Check out our [Resources](https://github.com/zmap/zmap/tree/main#resources) section for FAQ and feel free to chime in on [Github Discussions](https://github.com/zmap/zmap/discussions) if you're still unclear and we'll be happy to assist!

**Describe the bug**
ZMap intermittently ignores the specified `--rate` limit during scans. This issue can occur at any point during a scan and persists until the scan is completed.

Below is a screenshot illustrating this behavior.
<img width="510" alt="image" src="https://github.com/user-attachments/assets/f894adc3-44dc-4845-9c03-0ce491ac1112">
All 10 scanners in this example were invoked with `--rate=24000` (See the full command used below).

Unfortunately, no logs have been identified that indicate an issue.

**CLI Arguments**
Please paste your ZMap invocation below

```
/usr/local/sbin/zmap --quiet --target-port=XXX --rate=24000 --seed=9 --allowlist-file=/mnt/scanning/allowlist.conf --blocklist-file=/mnt/scanning/blocklist.conf --metadata-file=/mnt/scanning/scan.metadata.json --status-updates-file=/mnt/scanning/scan.status.json --probe-module=tcp_synscan --shards=10 --shard=3 --user-metadata={"scan_id":"xx", "scanner_id":"xx"} --sender-threads=1 --output-module=json --output-fields=timestamp_str,daddr,dport,saddr,sport,success --output-filter=success=1 && repeat=0
```

**Example Target IP**

Please list any IPs that can be used for testing behavior.

This bug seems to occur when running 

**Expected behavior**
The `--rate` limit should be respected for the entire duration of the scan.

**Environment:**
 - OS: Debian 12.8
 - Version: 4.2.0, also tested to build from HEAD (ae3fcbc) with same result.

**Additional context**
It may not be related but on occasion (approximately 5% of the time), scans fail to complete, and ZMap does not exit as expected. When this happens, logs repeatedly display the message: "Dropping X packets in the last second" roughly once every 5–10 seconds. If helpful, I can file a separate bug report for this issue. It seems similar to [#868](https://github.com/zmap/zmap/issues/868).

**Comment by phillip-stephens** [comment]
Hey @huxcrux, thanks for taking to the time to file this detailed issue. Let me take a look and see what's going on here.

**Comment by phillip-stephens** [comment]
Hey @huxcrux, took a look at this today and wasn't able to reproduce. I wanted to describe my setup and discuss how yours may be different and ask a few questions to help me troubleshoot.

On a single Ubuntu 24.04 VM, I ran 10 ZMap shards in parallel with:
```
sudo ./src/zmap --target-port=13579 --rate=24000 --seed=9 --metadata-file=meta.json --output-file=out --probe-module=tcp_synscan --shards=10 --shard=X --cores=X --user-metadata='{"scan_id":"xx", "scanner_id":"xx"}' --sender-threads=1 --output-module=json --output-fields=timestamp_str,daddr,dport,saddr,sport,success --output-filter=success=1 --status-updates-file=zmap-status-4.txt && repeat=0
```
I ran these for an hour and logged per-second the number of packets sent/received by the VM. Nothing else was running so ZMap's traffic should dominate the utilized traffic.

My logs show the overall mean traffic volume of the sum of the 10 invocations stayed very close to 240k (as it should). Additionally, the max tx packets is just 5,700 over the `--rate`, and this could be explained by the OS doing some background activity.

```
$ head Downloads/pps.log
2024-11-18 19:37:38;tx 241670 pkts/s;rx 11202 pkts/s; on interface enp0s8
2024-11-18 19:37:39;tx 241343 pkts/s;rx 10978 pkts/s; on interface enp0s8
...

~ ⌚ 22:28:23
$ cat Downloads/pps.log | cut -d ';' -f 2 | cut -d ' ' -f 2 | st
N	min	max	sum	mean	stddev
3610	3	245721	8.70426e+08	241115	7415.29
```

So had a few questions:

- Do you see this behavior every time you run this scan/how easily reproducible is it on your machine?
- Are your ZMap shards on separate VMs or all on the same one?
- How are you gathering your per-shard PPS?
- Do you see the same behavior with [v3.0.0](https://github.com/zmap/zmap/releases/tag/v3.0.0)

I think my next step will be to try to reproduce with Debian 12.8 and see if OS makes a difference.

**Comment by huxcrux** [comment]
> Hey @huxcrux, took a look at this today and wasn't able to reproduce. I wanted to describe my setup and discuss how yours may be different and ask a few questions to help me troubleshoot.
> 
> On a single Ubuntu 24.04 VM, I ran 10 ZMap shards in parallel with:
> 
> ```
> sudo ./src/zmap --target-port=13579 --rate=24000 --seed=9 --metadata-file=meta.json --output-file=out --probe-module=tcp_synscan --shards=10 --shard=X --cores=X --user-metadata='{"scan_id":"xx", "scanner_id":"xx"}' --sender-threads=1 --output-module=json --output-fields=timestamp_str,daddr,dport,saddr,sport,success --output-filter=success=1 --status-updates-file=zmap-status-4.txt && repeat=0
> ```
> 
> I ran these for an hour and logged per-second the number of packets sent/received by the VM. Nothing else was running so ZMap's traffic should dominate the utilized traffic.
> 
> My logs show the overall mean traffic volume of the sum of the 10 invocations stayed very close to 240k (as it should). Additionally, the max tx packets is just 5,700 over the `--rate`, and this could be explained by the OS doing some background activity.
> 
> ```
> $ head Downloads/pps.log
> 2024-11-18 19:37:38;tx 241670 pkts/s;rx 11202 pkts/s; on interface enp0s8
> 2024-11-18 19:37:39;tx 241343 pkts/s;rx 10978 pkts/s; on interface enp0s8
> ...
> 
> ~ ⌚ 22:28:23
> $ cat Downloads/pps.log | cut -d ';' -f 2 | cut -d ' ' -f 2 | st
> N	min	max	sum	mean	stddev
> 3610	3	245721	8.70426e+08	241115	7415.29
> ```
> 
> So had a few questions:
> 
>     * Do you see this behavior every time you run this scan/how easily reproducible is it on your machine?
> 
>     * Are your ZMap shards on separate VMs or all on the same one?
> 
>     * How are you gathering your per-shard PPS?
> 
>     * Do you see the same behavior with [v3.0.0](https://github.com/zmap/zmap/releases/tag/v3.0.0)
> 
> 
> I think my next step will be to try to reproduce with Debian 12.8 and see if OS makes a difference.

So had a few questions:

    * Do you see this behavior every time you run this scan/how easily reproducible is it on your machine?
    - Around 50-60% of my scans result in rate being ignored. Most of the times it occurs within the first 30 minutes however it could occur later in as well.

    * Are your ZMap shards on separate VMs or all on the same one?
    - I run ZMap on 10 different VMs, All VMs are KVM based and have an public IP on one of its interface.

    * How are you gathering your per-shard PPS?
    - I use Prometheus node-exporter for metrics on all scanner meaning the PPS reported includes other system services, however since the machines are dedicated scanners other traffic isn't a significant amount (when ZMap is not running the machines report around 5PPS with spikes to ~20).

    * Do you see the same behavior with [v3.0.0](https://github.com/zmap/zmap/releases/tag/v3.0.0)
    - I will test running ZMAP3 and keep you posted on the result.


I also have an update regarding the dropped logs. This was my scanners just being to slow writing to disk. I buffer the output in memory prior to writing to disk and I no longer see dropped packages meaning the only Issue we currently se is the rate being ignored.

Thanks alot for taking this bug report seriously and just let me know in case there is anything I can do helping the investigation.

**Comment by huxcrux** [comment]
Another update.

We tested downgrade and to isolate the zmap process (we usually run it together with our log shipper and zgrab). I was able to still trigger the issue every time. However when running version 3.0.0 we had around 10-15% of scans ignoring the rate limit. And when running 4.2.0 we have around 50%.

I found that the metrics written by ZMap during a run contains an average rate and duration I can confirm this is ZMap sending more packages than instructed.

We collect the files once a run is completed and therefore i do not have any from yesterdays testing (test where canceled mid run to cram in more tests). I do however have this data stored for all runs we have done during the last couple of months.

Example from a run on the 18th (2 days ago): https://gist.github.com/huxcrux/1c32c1831863513ee71d0706f24324c0

We will run some tests on Ubuntu 24.04 today since you mentioned it worked during your testing (we are currently reinstalling half of our scanner VMs and I will share our findings in a few hours.

**Comment by phillip-stephens** [comment]
This is great info, thank you. Very useful that in reproducing I can just trust the ZMap metadata. I can also try having more shards at once to give better chance at reproducing on my end. I'm a bit overloaded on another project so might be a few days before I can get back to this. One thing you can try as well to unblock yourself is using `-B` to set bandwidth limitations if you're just trying to limit bandwidth utilization by each scanner, in case that's useful to your use case.

**Comment by huxcrux** [comment]
I tested with Ubuntu 24.04 and see the same behavior there.

We have however seen a change in behavior when running on "dedicated" cores. Meaning we pin VMs to specific cores. I need to look more into this and will share more details during the upcoming days.

I can do a few tests tomorrow using the bandwidth limiter to se if we get another behavoiur.

**Comment by phillip-stephens** [comment]
Hey @huxcrux, wanted to give you an update on this.

I took a look at this today and after trying several things, just wasn't able to reproduce. Bit of a disclaimer, my setup is a little different: I'm using a single Ubuntu VM and pinning multiple ZMap invocations to different cores. 

1. I was wondering if either the `-B` bandwidth option or core pinning had uncovered any new behavior for you?
1. Also, wondering about the specs on these VMs/Hosts you're using (cores, RAM, etc.). Without core-pinning, it'd be reasonable if you were seeing _less_ send rate if there was contention for running on a given core. But I can't explain seeing _more._


Thanks and really appreciate being so helpful in this process of getting to the bottom of this.

**Comment by huxcrux** [comment]
> I've also added a couple debugging logs to this branch: [phillip/903-debugging-logs](https://github.com/zmap/zmap/tree/phillip/903-debugging-logs). This will add logs around the adaptive rate-limiting and print variables + exit if our rate spikes above 10x the specified `--rate`. Apologies for asking you to run custom code, but not sure how to reproduce on my side so this would be really helpful.
> 
> If you could, run to compile: `git pull && git checkout phillip/903-debugging-logs && cmake . && make -j4`
> 
> Run with - `sudo ./src/zmap -p 80 -r 24000 --verbosity=5 --log-file=log -T 1 --output-file=/dev/null --seed=9 --probe-module=tcp_synscan --shards=10 --shard=3`.
> 
> And then send me the `log` file of the machine that has a rate spike. I ran this test locally on my machine and the log file accumulated about `220 kB/min` so the log file shouldn't take up much space on your machine.
> 
> Thanks again!

Hello again,

I tested a couple of runs with your test branch and every time the rate limit is ignored the debug output is stopped.

For simplicity I removed `--log-file=log` to get all output to stdout. Here is the output from one of my runs: https://gist.github.com/huxcrux/87cc7ccbc52c2fc441db0ba50961af69

The same result seems to be the case when running with bandwidth limit instead of rate: https://gist.github.com/huxcrux/7a960ea11c4850a228af24a7ab2be268

It may be related or not however when dedicating cores to my VMs I se the problem less frequently. It goes from ignoring the rate-limit around 50% of times to maybe 1%.

I have tested a bit more with different kernels and Switching between Debian and Ubuntu and the change in behavior between different combinations seems to be withing error margin.

Also regarding the hosts we are currently running KVM based VMs. The scanner machines have 4 cores and 8 Gib of ram (I have tested to bump this to 8/16 just to make sure). We use virtio based networking and virtio scsi controller for storage. The underlying hosts are running REHL OS with Intel Xeon Platinum 81** CPUs. VMs are always assigned cores from the same CPU to avoid potential NUMA related problems.

When pinning cores meaning each machine gets "dedicated" cores the failure rate seems to drop to around 1-2% while running without CPU pinning we get around 50%.

**Comment by phillip-stephens** [comment]
Hey @huxcrux, thanks for being honestly amazing in assisting with the debugging process, those logs were invaluable. 

I'm almost positive I've found the bug and implemented a fix in [phillip/903-force-delay-nonzero](https://github.com/zmap/zmap/tree/phillip/903-force-delay-nonzero). PR is [here](https://github.com/zmap/zmap/pull/907).

I attached details of reproducing the same issue in the PR and after our internal review we can get this fix merged into `main`.
